### PR TITLE
C#: Include type parameters in MaD format for generics

### PR DESCRIPTION
**Pull request by hvitved**
This PR changes the MaD format used for generics, to include the names of the type parameters, in both generic types and generic methods. For example, instead of writing

```yml
extensions:
  - addsTo:
      pack: codeql/csharp-all
      extensible: summaryModel
      data:
        - ["System.Collections.Generic", "IList<>", True, "Insert", "(System.Int32,T)", "", "Argument[1]", "Argument[this].Element", "value", "manual"]
        - ["System.Linq", "Enumerable", False, "Select<,>", "(System.Collections.Generic.IEnumerable<TSource>,System.Func<TSource,System.Int32,TResult>)", "", "Argument[0].Element", "Argument[1].Parameter[0]", "value", "manual"]
```

we now instead write

```yml
extensions:
  - addsTo:
      pack: codeql/csharp-all
      extensible: summaryModel
      data:
        - ["System.Collections.Generic", "IList<T>", True, "Insert", "(System.Int32,T)", "", "Argument[1]", "Argument[this].Element", "value", "manual"]
        - ["System.Linq", "Enumerable", False, "Select<TSource,TResult>", "(System.Collections.Generic.IEnumerable<TSource>,System.Func<TSource,System.Int32,TResult>)", "", "Argument[0].Element", "Argument[1].Parameter[0]", "value", "manual"]
```

This aligns with [the official documentation](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.select?view=net-7.0#system-linq-enumerable-select-2(system-collections-generic-ienumerable((-0))-system-func((-0-1)))), and also makes it clear what type parameters in signatures refer to.

The PR is best reviewed commit-by-commit:
- Commit 1: Change how we print existing MaD rows to include type parameters.
- Commit 2: Update the model conversion queries to output existing MaD rows in the new format.
- Commit 3: Convert existing models using model converter.
- Commit 4: Change how we parse MaD rows to include type parameters.
- Commit 5: Manually update some MaD rows.
- Commit 6: Update model generator to output one file per namespace.
- Commit 7: Regenerate `dotnet/runtime` models using new format.
- Commit 8: Allow for explicit interface names in MaD consistency check.
- Commit 9: Change note.

### CODE REVIEW

File: csharp/ql/lib/ext/System.Data.Entity.model.yml [hunk 1]
```diff
@@ -20,3 +12,10 @@ extensions:
       - ["System.Data.Entity", "Database", False, "SqlQuery", "(System.Type,System.String,System.Object[])", "", "Argument[1]", "sql-injection", "manual"]
       - ["System.Data.Entity", "Database", False, "SqlQuery<>", "(System.String,System.Object[])", "", "Argument[0]", "sql-injection", "manual"]
       - ["System.Data.Entity", "DbSet", False, "SqlQuery", "(System.String,System.Object[])", "", "Argument[0]", "sql-injection", "manual"]
+  - addsTo:
+      pack: codeql/csharp-all
+      extensible: summaryModel
+    data:
+      - ["System.Data.Entity", "DbSet<>", False, "Add", "(TEntity)", "", "Argument[0]", "Argument[this].Element", "value", "manual"]
+      - ["System.Data.Entity", "DbSet<>", False, "AddRange", "(System.Collections.Generic.IEnumerable<TEntity>)", "", "Argument[0].WithElement", "Argument[this]", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Data.Entity.model.yml]: This must be an example of the fallback mechanism from "InterpretCallable" (this method doesn't exist in the database).
Have you checked whether all patterns like `<>` and `<,>` and so on has been removed from the Yaml files?

File: csharp/ql/lib/ext/System.Data.Entity.model.yml [hunk 2]
```diff
@@ -10,12 +10,12 @@ extensions:
...
+      - ["System.Data.Entity", "Database", False, "SqlQuery<TElement>", "(System.String,System.Object[])", "", "Argument[0]", "sql-injection", "manual"]
       - ["System.Data.Entity", "DbSet", False, "SqlQuery", "(System.String,System.Object[])", "", "Argument[0]", "sql-injection", "manual"]
   - addsTo:
       pack: codeql/csharp-all
       extensible: summaryModel
     data:
-      - ["System.Data.Entity", "DbSet<>", False, "Add", "(TEntity)", "", "Argument[0]", "Argument[this].Element", "value", "manual"]
-      - ["System.Data.Entity", "DbSet<>", False, "AddRange", "(System.Collections.Generic.IEnumerable<TEntity>)", "", "Argument[0].WithElement", "Argument[this]", "value", "manual"]
-      - ["System.Data.Entity", "DbSet<>", False, "Attach", "(TEntity)", "", "Argument[0]", "Argument[this].Element", "value", "manual"]
+      - ["System.Data.Entity", "DbSet<TEntity>", False, "Add", "(TEntity)", "", "Argument[0]", "Argument[this].Element", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Data.Entity.model.yml:19]: Ah, it looks like you also put in the manual conversion as well.

File: misc/scripts/models-as-data/generate_flow_model.py [hunk 3]
```diff
@@ -141,58 +127,57 @@ def runQuery(self, query):
...
-        return f"""# THIS FILE IS AN AUTO-GENERATED MODELS AS DATA FILE. DO NOT EDIT.
-# Definitions of models for the {self.friendlyname} framework.
-extensions:
-{sinkAddsTo}{sourceAddsTo}{summaryAddsTo}{neutralAddsTo}"""
+        return helpers.merge(summaryAddsTo, sinkAddsTo, sourceAddsTo, neutralAddsTo)
 
     def makeTypeBasedContent(self):
         if self.generateTypeBasedSummaries:
-            typeBasedSummaryAddsTo = self.getAddsTo("CaptureTypeBasedSummaryModels.ql", "extSummaryModel")
+            typeBasedSummaryAddsTo = self.getAddsTo("CaptureTypeBasedSummaryModels.ql", helpers.summaryModelPredicate)
```
Comment (michaelnebel) [inline misc/scripts/models-as-data/generate_flow_model.py:165]: Thank you for fixing this as well.

File: misc/scripts/models-as-data/generate_flow_model.py [hunk 4]
```diff
@@ -141,58 +127,57 @@ def runQuery(self, query):
...
-        with open(target, "w") as targetYml:
-            targetYml.write(content)
-        print("Models as data extensions written to " + target)
+    def save(self, extensions, extension):
+        # Create a file for each namespace and save models.
+        extensionTemplate = """# THIS FILE IS AN AUTO-GENERATED MODELS AS DATA FILE. DO NOT EDIT.
+extensions:
+{0}"""
+        for entry in extensions:
+            target = os.path.join(self.generatedFrameworks, entry + extension)
```
Comment (michaelnebel) [inline misc/scripts/models-as-data/generate_flow_model.py:177]: Did you also try running this for java?

**Comment by michaelnebel** [comment]
- Do you know which summaries in `System` that disappeared?
- Will you start a DCA run?

File: csharp/ql/lib/ext/System.Collections.Generic.model.yml [hunk 5]
```diff
@@ -66,8 +66,8 @@ extensions:
       - ["System.Collections.Generic", "SortedDictionary<TKey,TValue>", False, "get_Values", "()", "", "Argument[this].Element.Property[System.Collections.Generic.KeyValuePair<,>.Value]", "ReturnValue.Element", "value", "manual"]
       - ["System.Collections.Generic", "SortedDictionary<TKey,TValue>+KeyCollection", False, "GetEnumerator", "()", "", "Argument[this].Element", "ReturnValue.Property[System.Collections.Generic.SortedDictionary<,>+KeyCollection+Enumerator.Current]", "value", "manual"]
       - ["System.Collections.Generic", "SortedDictionary<TKey,TValue>+ValueCollection", False, "GetEnumerator", "()", "", "Argument[this].Element", "ReturnValue.Property[System.Collections.Generic.SortedDictionary<,>+ValueCollection+Enumerator.Current]", "value", "manual"]
-      - ["System.Collections.Generic", "SortedList<,>+KeyList", False, "Clear", "()", "", "Argument[this].WithoutElement", "Argument[this]", "value", "manual"]
```
Comment (michaelnebel) [inline csharp/ql/lib/ext/System.Collections.Generic.model.yml:78]: Have you searched all the yaml files using the regex `<,*>`?

File: misc/scripts/models-as-data/generate_flow_model.py (hunk 4 above)
Comment (hvitved) [reply misc/scripts/models-as-data/generate_flow_model.py:177]: Nope :-)

**Comment by hvitved** [comment]
> * Do you know which summaries in `System` that disappeared?

That would be the auto-generated summaries from https://github.com/github/codeql/pull/14662/commits/136adb2df24a96702d11c75aa4fc07265a461763.

> * Will you start a DCA run?

Already did :-)

File: misc/scripts/models-as-data/generate_flow_model.py (hunk 4 above)
Comment (michaelnebel) [reply misc/scripts/models-as-data/generate_flow_model.py:177]: If you are interested then you can run modify the `mad_modelDiff.yml` workflow file to also be triggered on the shared python code.

Comment (hvitved) [reply misc/scripts/models-as-data/generate_flow_model.py:177]: I don't know how to run that on my PR branch, do you?

Comment (michaelnebel) [reply misc/scripts/models-as-data/generate_flow_model.py:177]: Can't you just change the paths section of the workflow file? If that doesn't work - then I don't know.

**Review by michaelnebel — APPROVED** [review]
Thank you very much for doing this! 👍

**Pull request merged**